Refactor Connection.js, transports, and Toolkit.js

abmusse opened the pull request:
## Connection class:

Constructor now supports passing a single object for configuration. see issue #25

Old signature with positional arguments are still supported.

reworked Connection.run() method:

run function invokes the transport function and passes:

- config: the object with configuration details
- xmlInput: the input xml document
- done: function to be invoked by the transport

run function now provides error first callback by default this resolved issue #2. 

This can be configured through `returnError` property.

removed setTimeout() and timeout property since we no longer have sync mode, see issue #32 

renamed internal variables away from I_* prefix that was used before.

No longer set defaults for transport options simply pass the options object to the transport.

Transports now deals with setting defaults and validate required parameters are OK.

renamed `getConnection()` to `getTransportOptions()`

## Transports
- updated to interface to accept config, xmlInput, done

- In `istoredp.js` moved to use `iPLUGR512K` procedure which returns a result set.

- transports were moved lib/transports/ directory

## Toolkit.js 
Updated Toolkit methods to return error first callbacks

kadler commented:
Why was a new PR opened instead of updating the existing one?

abmusse commented:
The original PR contained more commits, updates to the tests and other files.

I decided to exclude those commits from this PR and have those changes be in a separate PR next.

Also these changes were added to separate branch therefore could not use the same PR.

If its easier I can push these commits over to `reworkConnection` branch and reopen the previous PR.

kadler commented on lib/Connection.js:
```diff
@@ -16,82 +16,71 @@
+      options.username = username;
+      options.password = password;
+      options.transport = 'idb';
+      options.returnError = false;
```
could do something like:
```js
options = {
  database: optionsOrDb,
  username: username,
  ...
}

abmusse replied:
Will patch to be

```javascript
      options = {
        returnError: false,
        transport: 'idb',
        transportOptions: {
          database: optionsOrDb,
          username,
          password
        }
      }
```

kadler commented on lib/Connection.js:
```diff
@@ -16,82 +16,71 @@
+      options = {};
+      options.database = optionsOrDb;
+      options.username = username;
+      options.password = password;
```
are `database`, `username`, and `password` not transport options?

abmusse replied:
Yes will add those properties to `options.transportOptions` sub object.

kadler commented on lib/Connection.js:
```diff
@@ -16,82 +16,71 @@
+    delete options.transport;
+    delete options.returnError;
+
+    this.transportOptions = { ...options };
```
Why is the user not passing in a `transportOptions` sub object?

abmusse replied:
I was thinking the user would pass in the object in the format like:
```javascript
{
 returnError,
 transport,
 username,
 password,
....
}
```

We could design it for the user to pass in the object like:

```javascript
{
   returnError,
   transport,
   transportOptions: { username, password, ....}
}
```

kadler replied:
I think that's best. Any transport-specific options are in the `transportOptions` key and you just pass it through to the transport. Then there's no confusion about what key is used for what.

abmusse commented:
@kadler pushed up an update  https://github.com/IBM/nodejs-itoolkit/pull/42/commits/d041d9949d9e0079e447017d8fda7ee7a2e5ccde that uses:

```javascript
options = {
   returnError,
   transport,
   transportOptions: { username, password, ....}
}

```
Should resolve the above suggestions

abmusse commented:
> LGTM

I'm going to clean up a bit and squash incremental commits together to be more stream-lined

abmusse commented:
Cleaned up and Squashed:

![image](https://user-images.githubusercontent.com/33973272/54150137-c3ae0e00-4405-11e9-8ac5-c4782bb6357f.png)

Into:

![image](https://user-images.githubusercontent.com/33973272/54150190-d7f20b00-4405-11e9-841b-f0afa0d6f7ba.png)


Going merge PR into v1.0-dev
